### DISCUSSION TITLE
dep-type msbuild not supporting custom registries (from NuGet.Config)?

### DISCUSSION THREAD

**Discussion by japj** [renovatebot/renovate#13396] [Request Help]
Hello,

I ran into an issue where renovate is not able to update a private msbuild-sdk version in the global.json (I get "Failed to look up dependency " messages).

As far as I can tell the dependencies from `extractMsbuildGlobalManifest` in https://github.com/renovatebot/renovate/blob/main/lib/manager/nuget/extract/global-manifest.ts are not extended with information on `registryUrls` like `extractPackageFile` in https://github.com/renovatebot/renovate/blob/main/lib/manager/nuget/extract.ts does for others.

the code currently does:
```ts
  if (packageFile.endsWith('global.json')) {
    return extractMsbuildGlobalManifest(content, packageFile);
  }
```

while other code adds the registryUrls information:
```ts
   deps = extractDepsFromXml(parsedXml).map((dep) => ({
      ...dep,
      ...(registryUrls && { registryUrls }),
    }));
```

and
```ts
 if (registryUrls) {
        dep.registryUrls = registryUrls;
 }
```

Is this something I can "workaround" from my config.js or would this require a fix in renovate?

**Comment by rarkins** [comment]
You can customize registryUrls using packageRules

**Reply by japj** [reply]
Hmm, this does not seem to work yet for me. Perhaps I configured it incorrectly?
I added several registryUrls to the packageRules for DepType msbuild-sdk

```ts
"packageRules": [{
    "matchDepType": ["msbuild-sdk"],
    "registryUrls": [ "on-premise AzureDevOps feed" ]
}]
```

I see this also in the `Combined config` logging, but the actual package still cannot be found:

```txt
                packageFile": "global.json",
                "deps": [
                  {
                    "depType": "msbuild-sdk",
                    "depName": "CanonPP.Common.CodeGenerator",
                    "currentValue": "4.0.0",
                    "datasource": "nuget",
                    "depIndex": 1,
                    "updates": [],
                    "warnings": [
                      {
                        "topic": "CanonPP.Common.CodeGenerator",
                        "message": "Failed to look up dependency CanonPP.Common.CodeGenerator"
                      }
                    ],
                    "versioning": "nuget"
                  },
```

If I compare this with a regular nuget package I see registryUrls also in the deps like:

```txt
                  {
                   "datasource": "nuget",
                    "depType": "nuget",
                    "depName": "CanonPP.Interfaces.DataDrivenForms",
                    "currentValue": "0.26.0",
                    "registryUrls": [
                      "https://api.nuget.org/v3/index.json",
                      etc...
```

Trying this workaround route also makes me wonder if it is the right path, since I am now duplicating information that is already contained in the NuGet.config. It is just that this NuGet.config information does not seem to be used yet by the `msbuild-sdk` depType package detection.
